src/core/bst.h

```diff
@@ -188,361 +133,540 @@ class BST {
+    virtual void accept_stmt(StmtVisitor* v);
 
-    BST_Index() : BST_slice(BST_TYPE::Index) {}
+    BST_DeleteAttr() : BST_stmt(BST_TYPE::DeleteAttr) {}
 
-    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::Index;
+    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::DeleteAttr;
 };
 
-class BST_keyword : public BST {
+class BST_DeleteName : public BST_stmt {
 public:
-    // no lineno attributes
-    BST_expr* value;
-    InternedString arg;
+    InternedString id;
+    ScopeInfo::VarScopeType lookup_type;
+    int vreg = VREG_UNDEFINED;
+
+    // Only valid for lookup_type == DEREF:
+    DerefInfo deref_info = DerefInfo({ INT_MAX, INT_MAX });
+    // Only valid for lookup_type == CLOSURE:
+    int closure_offset = -1;
+
 
     virtual void accept(BSTVisitor* v);
+    virtual void accept_stmt(StmtVisitor* v);
 
-    BST_keyword() : BST(BST_TYPE::keyword) {}
+    BST_DeleteName() : BST_stmt(BST_TYPE::DeleteName) {}
 
-    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::keyword;
+    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::DeleteName;
 };
 
-class BST_List : public BST_expr {
+class BST_DeleteSub : public BST_stmt {
 public:
-    std::vector<BST_expr*> elts;
-    AST_TYPE::AST_TYPE ctx_type;
+    int vreg_value = VREG_UNDEFINED;
+    int vreg_slice = VREG_UNDEFINED;
 
     virtual void accept(BSTVisitor* v);
-    virtual void* accept_expr(ExprVisitor* v);
+    virtual void accept_stmt(StmtVisitor* v);
 
-    BST_List() : BST_expr(BST_TYPE::List) {}
+    BST_DeleteSub() : BST_stmt(BST_TYPE::DeleteSub) {}
 
-    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::List;
+    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::DeleteSub;
 };
 
-class BST_Name : public BST_expr {
+class BST_DeleteSubSlice : public BST_stmt {
 public:
-    AST_TYPE::AST_TYPE ctx_type;
-    InternedString id;
+    int vreg_value = VREG_UNDEFINED;
+    int vreg_lower = VREG_UNDEFINED;
+    int vreg_upper = VREG_UNDEFINED;
 
-    // The resolved scope of this name.  Kind of hacky to be storing it in the BST node;
-    // in CPython it ends up getting "cached" by being translated into one of a number of
-    // different bytecodes.
-    ScopeInfo::VarScopeType lookup_type;
+    virtual void accept(BSTVisitor* v);
+    virtual void accept_stmt(StmtVisitor* v);
 
-    // These are only valid for lookup_type == FAST or CLOSURE
-    // The interpreter and baseline JIT store variables with FAST and CLOSURE scopes in an array (vregs) this specifies
-    // the zero based index of this variable inside the vregs array. If uninitialized it's value is -1.
-    int vreg;
-    bool is_kill = false;
+    BST_DeleteSubSlice() : BST_stmt(BST_TYPE::DeleteSubSlice) {}
 
-    // Only valid for lookup_type == DEREF:
-    DerefInfo deref_info = DerefInfo({ INT_MAX, INT_MAX });
-    // Only valid for lookup_type == CLOSURE:
-    int closure_offset = -1;
+    static const BST_TYPE::BST_TYPE TYPE = BST_TYPE::DeleteSubSlice;
+};
 
+class BST_Ellipsis : public BST_dst {
```

Comment:
I think this can be treated as a constant rather than another node (looks like this is how CPython does it).